compiler/optimizer/LoopCanonicalizer.cpp

```diff
@@ -1837,7 +1837,74 @@ void TR_LoopCanonicalizer::canonicalizeDoWhileLoop(TR_RegionStructure *doWhileLo
    return;
    }
 
+/**
+ * \brief
+ *    Determines whether the node is a trivial (arithmetic and checks only) array operation.
+ *
+ * \param visited
+ *    The list of visited descendants which should not be examined.
+ *
+ * \param node
+ *    The node to examine.
+ *
+ * \return
+ *    \c true if \p node is a trivial array operation; \c false otherwise.
+ *
+ * \details
+ *    This API is used to identify trivial array operations while performing induction variable elimination. In cases
+ *    where the array operation is considered trivial there may be a negative trade-off between eliminating an extra
+ *    induction variable compared to leaving it present. The typical scenario this can be encountered in is an array
+ *    copy operation which may look like this (in pseudocode):
+ *
+ *    ```
+ *    while (j < x) {
+ *        a[i++] = b[j++];
+ *    }
+ *    ```
+ *
+ *    If induction variables `i` and `j` are not known to be constants outside the loop then eliminating one induction
+ *    variable forces us to rewrite the above loop to be:
+ *
+ *    ```
+ *    temp = i - j;
+ *    while (j < x) {
+ *        a[temp + j] = b[j++]
+ *    }
+ *    i = temp + j;
+ *    ```
+ *
+ *    i.e. we have to rewrite one induction variable in terms of the other. The latter loop is more complex than the
+ *    former because the recalculation of the induction variable now dominates the computation within the loop and
+ *    hence it is not a worth while trade-off. Performing such an induction variable elimination can result in degraded
+ *    performance which could then be amplified by loop unrolling at later stages of the optimization strategy.
+ */
+static bool isTrivialArrayOperation(TR::NodeChecklist &visited,TR::Node *node)
```

Comment:
given this check code will cause us to fail to do something I think this needs some more trace to explain which tree and why we gave up on a particular operation. Debugging from only a log is going to be pretty hard without some more trace IMO.

Reply:
I have added some tracing in the latest commit. Let me know if you would like to see more tracing